Make script::Method a value type

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #21853 Add interface declarations to JIT
* #21874 Add slot_list/slot_iterator
* #21814 script::Module as a view.
* **#21675 Make script::Method a value type**

Now that modules are always first class, a Method is simply a pair
of Function and Method object. It is no longer necessary to have
a reified table of Methods and removes the need for locking.

Differential Revision: [D15777725](https://our.internmc.facebook.com/intern/diff/D15777725)

zdevito committed:
Make script::Method a value type

Now that modules are always first class, a Method is simply a pair
of Function and Method object. It is no longer necessary to have
a reified table of Methods and removes the need for locking.

zdevito committed:
Update on "Make script::Method a value type"

Make script::Method a value type

Now that modules are always first class, a Method is simply a pair
of Function and Method object. It is no longer necessary to have
a reified table of Methods and removes the need for locking.

gh-metadata: pytorch pytorch 21675 gh/zdevito/64/head

zdevito committed:
Update on "Make script::Method a value type"

Make script::Method a value type

Now that modules are always first class, a Method is simply a pair
of Function and Method object. It is no longer necessary to have
a reified table of Methods and removes the need for locking.

gh-metadata: pytorch pytorch 21675 gh/zdevito/64/head

zdevito committed:
Update on "Make script::Method a value type"

Make script::Method a value type

Now that modules are always first class, a Method is simply a pair
of Function and Method object. It is no longer necessary to have
a reified table of Methods and removes the need for locking.

gh-metadata: pytorch pytorch 21675 gh/zdevito/64/head

zdevito committed:
Update on "Make script::Method a value type"

Make script::Method a value type

Now that modules are always first class, a Method is simply a pair
of Function and Method object. It is no longer necessary to have
a reified table of Methods and removes the need for locking.

gh-metadata: pytorch pytorch 21675 gh/zdevito/64/head

suo commented on torch/csrc/jit/script/module.h:
```diff
@@ -215,15 +213,11 @@ struct TORCH_API Module {
-    // can be removed as well
-    get_offset(name, EntityType::METHOD);
-    AT_ERROR("unreachable");
+    AT_ERROR("Method '", name, "' is not defined.");
```
Let's swap this out for one of the new error methods while we're here

suo commented on torch/csrc/jit/script/module.h:
```diff
@@ -239,13 +233,11 @@ struct TORCH_API Module {
+  const std::vector<Method> get_methods() const {
+    return fmap(class_compilation_unit().get_functions(),
+                [&](const std::shared_ptr<Function> &func) {
+                  return Method(const_cast<Module *>(this), func);
```
Can we remove this const_cast? It's not clear to me why methods need a mutable reference to their owner anyway

zdevito committed:
Update on "Make script::Method a value type"

Make script::Method a value type

Now that modules are always first class, a Method is simply a pair
of Function and Method object. It is no longer necessary to have
a reified table of Methods and removes the need for locking.

gh-metadata: pytorch pytorch 21675 gh/zdevito/64/head

zdevito committed:
Update on "Make script::Method a value type"

Make script::Method a value type

Now that modules are always first class, a Method is simply a pair
of Function and Method object. It is no longer necessary to have
a reified table of Methods and removes the need for locking.

gh-metadata: pytorch pytorch 21675 gh/zdevito/64/head

zdevito committed:
Update on "Make script::Method a value type"

Make script::Method a value type

Now that modules are always first class, a Method is simply a pair
of Function and Method object. It is no longer necessary to have
a reified table of Methods and removes the need for locking.

gh-metadata: pytorch pytorch 21675 gh/zdevito/64/head

zdevito committed:
Update on "Make script::Method a value type"

Make script::Method a value type

Now that modules are always first class, a Method is simply a pair
of Function and Method object. It is no longer necessary to have
a reified table of Methods and removes the need for locking.

gh-metadata: pytorch pytorch 21675 gh/zdevito/64/head

zdevito committed:
Update on "Make script::Method a value type"

Make script::Method a value type

Now that modules are always first class, a Method is simply a pair
of Function and Method object. It is no longer necessary to have
a reified table of Methods and removes the need for locking.

gh-metadata: pytorch pytorch 21675 gh/zdevito/64/head

zdevito committed:
Update on "Make script::Method a value type"

Make script::Method a value type

Now that modules are always first class, a Method is simply a pair
of Function and Method object. It is no longer necessary to have
a reified table of Methods and removes the need for locking.

gh-metadata: pytorch pytorch 21675 gh/zdevito/64/head

zdevito committed:
Update on "Make script::Method a value type"

Make script::Method a value type

Now that modules are always first class, a Method is simply a pair
of Function and Method object. It is no longer necessary to have
a reified table of Methods and removes the need for locking.

gh-metadata: pytorch pytorch 21675 gh/zdevito/64/head

zdevito committed:
Update on "Make script::Method a value type"

Make script::Method a value type

Now that modules are always first class, a Method is simply a pair
of Function and Method object. It is no longer necessary to have
a reified table of Methods and removes the need for locking.

gh-metadata: pytorch pytorch 21675 gh/zdevito/64/head
